### DISCUSSION TITLE
[AttributeRuler] Remove a pattern

### DISCUSSION THREAD

**Discussion by databill86** [explosion/spaCy#8888] [Help: Coding & Implementations] [answered]
Hello, 
I would like to delete a pattern from the attribute rule, while the model is already instantiated, but it seems that [there is no way to do this](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/attributeruler.py).
Also, it seems that the ID for of a pattern:

```
[
  {
    "patterns": [[{"TAG": "VB"}]], 
    "attrs": {"POS": "VERB"}
  },
 ....
]
```

[is defined as ](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/attributeruler.py#L200): 

```
key = str(len(self.attrs))
self.matcher.add(self.vocab.strings.add(key), patterns)
```

We can't delete a specific pattern because we might end up deleting another pattern having the same "attrs". 

The only solution I've found is to reload the AttributeRuler pipeline. 

Is there something that I'm missing ? 

Thanks

**Comment by adrianeboyd** [comment, marked as answer]
No, the patterns are just numbered sequentially using `len(self.attrs)` as they're added because they don't have any other kinds of labels or names.

If you can identify the number of the pattern you want to remove, I think you should be able to remove it just for the currently loaded pipeline by removing it directly from the internal matcher with `self.matcher.remove(str(number))`. The patterns are numbered starting at 0, and the key is the string of the number rather than an int.

If you save and reload the pipeline, the pattern will be reloaded again, so this is really just a local hack and not a real remove. And I haven't tested this myself, but I think if the matcher doesn't return the pattern the other saved settings will just be ignored when you call the attribute ruler.

In general, it would be cleaner to remove the attribute ruler and add a new attribute ruler in the same position in the pipeline with the patterns that you want.

```python
patterns = nlp.get_pipe("attribute_ruler").patterns
# remove the first pattern
patterns = patterns[1:]
nlp.remove_pipe("attribute_ruler")
ar = nlp.add_pipe("attribute_ruler", before="lemmatizer")
ar.add_patterns(patterns)
```

**Reply by databill86** [reply]
Thank you for your answer. 

> In general, it would be cleaner to remove the attribute ruler and add a new attribute ruler in the same position in the pipeline with the patterns that you want.

That's what I thought. Actually I have created a custom attribute rule that come after the lemmatizer. So I just have to reload it as you mentioned. 

Thank you.
